bin/rebuild-package.sh

```diff
@@ -0,0 +1,311 @@
+#!/bin/bash
+
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+###
+### The following environment variables are expected to be set in order for
+### the various elements in this script to function properly.  Listed first
+### are two defaults we rely on from gitlab, then three we set up in the
+### variables section of gitlab ourselves, and finally four variables
+### written into the .gitlab-ci.yml file.
+###
+### CI_PROJECT_DIR
+### CI_JOB_NAME
+###
+### AWS_ACCESS_KEY_ID
+### AWS_SECRET_ACCESS_KEY
+### SPACK_SIGNING_KEY
+###
+### CDASH_BASE_URL
+### ROOT_SPEC
+### DEPENDENCIES
+### MIRROR_URL
+###
+
+shopt -s expand_aliases
+
+export FORCE_UNSAFE_CONFIGURE=1
+
+TEMP_DIR="${CI_PROJECT_DIR}/jobs_scratch_dir"
+
+JOB_LOG_DIR="${TEMP_DIR}/logs"
+SPEC_DIR="${TEMP_DIR}/specs"
+LOCAL_MIRROR="${CI_PROJECT_DIR}/local_mirror"
+BUILD_CACHE_DIR="${LOCAL_MIRROR}/build_cache"
+SPACK_BIN_DIR="${CI_PROJECT_DIR}/bin"
+CDASH_UPLOAD_URL="${CDASH_BASE_URL}/submit.php?project=Spack"
+DEP_JOB_RELATEBUILDS_URL="${CDASH_BASE_URL}/api/v1/relateBuilds.php"
+declare -a JOB_DEPS_PKG_NAMES
+
+export SPACK_ROOT=${CI_PROJECT_DIR}
+export PATH="${SPACK_BIN_DIR}:${PATH}"
+export GNUPGHOME="${CI_PROJECT_DIR}/opt/spack/gpg"
+
+mkdir -p ${JOB_LOG_DIR}
+mkdir -p ${SPEC_DIR}
+
+cleanup() {
+    set +x
+
+    if [ -z "$exit_code" ] ; then
+
+        exit_code=$1
+        if [ -z "$exit_code" ] ; then
+            exit_code=0
+        fi
+
+        restore_io
+
+        if [ "$( type -t finalize )" '=' 'function' ] ; then
+            finalize "$JOB_LOG_DIR/cdash_log.txt"
+        fi
+
+        # We can clean these out later on, once we have a good sense for
+        # how the logging infrastructure is working
+        # rm -rf "$JOB_LOG_DIR"
+    fi
+
+    \exit $exit_code
+}
+
+alias exit='cleanup'
+
+begin_logging() {
+    trap "cleanup 1; \\exit \$exit_code" INT TERM QUIT
+    trap "cleanup 0; \\exit \$exit_code" EXIT
+
+    rm -rf "$JOB_LOG_DIR/cdash_log.txt"
+
+    # NOTE: Here, some redirects are set up
+    exec 3>&1 # fd 3 is now a dup of stdout
+    exec 4>&2 # fd 4 is now a dup of stderr
+
+    # stdout and stderr are joined and redirected to the log
+    exec &> "$JOB_LOG_DIR/cdash_log.txt"
+
+    set -x
+}
+
+restore_io() {
+    exec  >&-
+    exec 2>&-
+
+    exec  >&3
+    exec 2>&4
+
+    exec 3>&-
+    exec 4>&-
+}
+
+finalize() {
+    # If you define a finalize function:
+    #  - it will always be called at the very end of the script
+    #  - the log file will be passed in as the first argument, and
+    #  - the code in this function will not be logged.
+    echo "The full log file is located at $1"
+    # TODO: send this log data to cdash!
```

Comment:
Is this still a TODO? Is this pending some other thing being done?

Reply:
I have a refactoring pass planned for the rebuild script.  One of the things I want to do there is to take control of job registration with CDash (instead of letting `spack install ...` do it).  This way, any errors encountered in the bash script can be potentially included in the information given to CDash and available for viewing there.  As it stands, nothing which may go wrong before the `spack install...` command could possibly be communicated to CDash.  I think it would also be nice if the full output log produced by this rebuild script is pushed to CDash at that point as a "note".  But I thought we wanted to get this topic merged as quickly as possible, so I didn't want to hold it up with a bunch of things that could be done as follow-on topics, hence that `TODO` comment.